js/src/components/google-ads-account-card/google-ads-account-card.js

```diff
@@ -1,37 +1,116 @@
 import useGoogleAdsAccount from '.~/hooks/useGoogleAdsAccount';
 import ConnectedGoogleAdsAccountCard from './connected-google-ads-account-card';
 import NonConnected from './non-connected';
 import AuthorizeAds from './authorize-ads';
-import { GOOGLE_ADS_ACCOUNT_STATUS } from '.~/constants';
+import ClaimAccount from './claim-account';
+import ClaimAccountModal from './claim-account-modal';
+import SpinnerCard from '../spinner-card';
+import useGoogleAdsAccountStatus from '.~/hooks/useGoogleAdsAccountStatus';
+import usePrevious from '.~/hooks/usePrevious';
+import useFetchCreateAdsAccount from '.~/hooks/useFetchCreateAdsAccount';
+import useDispatchCoreNotices from '.~/hooks/useDispatchCoreNotices';
 
 export default function GoogleAdsAccountCard() {
+	const [ claimModalOpen, setClaimModalOpen ] = useState( false );
 	const { google, scope } = useGoogleAccount();
-	const { googleAdsAccount } = useGoogleAdsAccount();
+	const {
+		googleAdsAccount,
+		refetchGoogleAdsAccount,
+		hasFinishedResolution: hasResolvedGoogleAdsAccount,
+	} = useGoogleAdsAccount();
+	const { hasAccess, hasFinishedResolution } = useGoogleAdsAccountStatus();
+	const [ fetchCreateAdsAccount, { loading } ] = useFetchCreateAdsAccount();
+	const { createNotice } = useDispatchCoreNotices();
+	const previousHasAccess = usePrevious( hasAccess );
+
+	const handleOnCreateAccount = useCallback( () => {
+		setClaimModalOpen( true );
+	}, [] );
+
+	const handleOnRequestClose = useCallback( () => {
+		setClaimModalOpen( false );
+	}, [] );
+
+	useEffect( () => {
+		const checkAccessChange = async () => {
+			// Access has changed, continue setup process
+			if ( hasAccess === true && previousHasAccess === false ) {
+				try {
+					await fetchCreateAdsAccount();
+				} catch ( error ) {
+					if (
+						error.status !== 428 &&
+						error?.billing_status !==
+							GOOGLE_ADS_BILLING_STATUS.PENDING
+					) {
+						createNotice(
+							'error',
+							__(
+								'Unable to connect Google Ads account. Please try again later.',
+								'google-listings-and-ads'
+							)
+						);
+					}
+				}
+
+				await refetchGoogleAdsAccount();
+			}
+		};
+
+		checkAccessChange();
+	} );
+
+	if ( ! hasFinishedResolution ) {
+		return <SpinnerCard />;
+	}
 
 	if ( ! google || ! googleAdsAccount ) {
-		return <NonConnected />;
+		return <NonConnected onCreateAccount={ handleOnCreateAccount } />;
 	}
 
 	if ( ! scope.adsRequired ) {
 		return <AuthorizeAds additionalScopeEmail={ google.email } />;
 	}
 
 	if ( googleAdsAccount.status === GOOGLE_ADS_ACCOUNT_STATUS.DISCONNECTED ) {
-		return <NonConnected />;
+		return <NonConnected onCreateAccount={ handleOnCreateAccount } />;
+	}
+
+	// Ads account has been created but we don't have access yet.
+	if ( googleAdsAccount.id && hasAccess === false ) {
+		return (
+			<Fragment>
+				{ claimModalOpen && (
+					<ClaimAccountModal
+						onRequestClose={ handleOnRequestClose }
+					/>
+				) }
```

Comment:
@eason9487 The thing here is that all the sub components within `GoogleAdsAccountCard` will be mounted and unmounted depending on various conditions. In the case of the modal, if we move the logic to display the modal to `CreateAccount` (which I did), it will not work since we are setting a state variable to display/hide the modal and when the component gets unmounted and mounted again, the state is reinitialized. That's why I originally had the modal state within `GoogleAdsAccountCard` since it won't be unmounted. I can move the state variable back but once again we'll have to resort to prop drilling to modify the variable. Right now the modal is not working, unless you have a better solution? 😁

Reply:
👍